QEfficient/peft/auto.py

```diff
@@ -0,0 +1,436 @@
+import logging
+import shutil
+import subprocess
+import warnings
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+import onnx
+import torch
+from peft import AutoPeftModelForCausalLM, load_peft_weights
+from torch import nn
+from transformers import GenerationConfig, StoppingCriteria, StoppingCriteriaList
+from transformers.generation.streamers import BaseStreamer
+
+from QEfficient.base.modeling_qeff import QEFFBaseModel
+from QEfficient.base.onnx_transforms import FP16ClipTransform, OnnxTransform, SplitTensorsTransform
+from QEfficient.base.pytorch_transforms import PytorchTransform
+from QEfficient.generation.cloud_infer import QAICInferenceSession
+from QEfficient.peft.onnx_transforms import AdapterWeightsToInputsTransform
+from QEfficient.peft.pytorch_transforms import PeftModelInputsTransform
+from QEfficient.transformers.pytorch_transforms import CustomOpsTransform, KVCacheTransform
+from QEfficient.utils._utils import get_padding_shape_from_config
+from QEfficient.utils.cache import QEFF_HOME, to_hashable
+
+logger = logging.getLogger(__name__)
+
+
+class QEffAutoPeftModelForCausalLM(QEFFBaseModel):
+    _pytorch_transforms: List[PytorchTransform] = [CustomOpsTransform, KVCacheTransform, PeftModelInputsTransform]
+    _onnx_transforms: List[OnnxTransform] = [FP16ClipTransform, AdapterWeightsToInputsTransform, SplitTensorsTransform]
+    _hf_auto_class = AutoPeftModelForCausalLM
+
+    @classmethod
+    def transform_names(cls) -> List[str]:
+        # Base class
+        return [x.__name__ for x in cls._pytorch_transforms + cls._onnx_transforms]
+
+    def __init__(self, model: nn.Module):
+        if model.active_peft_config.peft_type != "LORA":
+            raise NotImplementedError("Only LoRA models are supported")
+
+        super().__init__(model)
+
+        self.num_layers = model.config.num_hidden_layers
+        self.adapter_weights = {}
+        self.transform()
+
+    @property
+    def model_name(self) -> str:
+        return self.model.get_base_model().__class__.__name__ + "-lora"
+
+    @property
+    def model_hash(self) -> str:
+        # NOTE: model_config.to_diff_dict() has "_name_or_path" attribute which is the model card name or path.
+        # Using same card name will result in same hash. But, using a relative path for one run and
+        # absolute path for another run will result in different hash.
+        # The added complexity to resolve different paths to same location is not worth pursuing.
+        # Instead, advise the user to always provide same relative paths or absolute paths for local models.
+
+        # Compute the hash with: model_config, peft_config, transforms
+        mhash = hashlib.sha256()
+        mhash.update(to_hashable(self.model.get_base_model().config.to_diff_dict()))
+        mhash.update(to_hashable(self.model.active_peft_config.to_dict()))
+        mhash.update(to_hashable(self.transform_names()))
+        mhash = mhash.hexdigest()[:16]
```

Comment:
Should we use `SHAKE_128`? for variable length `hexdigest`? or `xxhash` -> faster

Reply:
I researched a bit and thought of using SHAKE, but it may not be supported in all platforms. hashlib uses OpenSSL system libraries to provide the support for hashing. In older versions of OpenSSL, it may not be supported. This is mentioned in python library documentation: https://docs.python.org/3/library/hashlib.html 

> Changed in version 3.9: Hashlib now uses SHA3 and SHAKE from OpenSSL if it provides it.

As for xxhash, it's adding unnecessary dependency for such a very small hash. We're not hashing GB's or TB's of data to go for xxhash to make it faster. So went with much simpler and standard way of doing things.

Reply:
Can we get rid of only considering first 16 characters of `hexdigest`? That's the cause of concern

Reply:
Even for large git repos (which use SHA256 hash) like linux kernel (millions of commits), prefix size of 11 is sufficient to avoid collision: https://stackoverflow.com/questions/32405922/in-my-repo-how-long-must-the-longest-hash-prefix-be-to-prevent-any-overlap